### PR TITLE
DocType: Mediated Order

### PR DESCRIPTION
**Pull request by cp-ps**

### CODE REVIEW

File: backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/order/createFrom/po_from_so/impl/CreatePOLineFromSOLinesAggregator.java [hunk 1]
```diff
@@ -160,11 +174,18 @@ private void copyUserIdFromSalesToPurchaseOrderLine(
...
 		for (final I_C_OrderLine salesOrderLine : purchaseOrderLine2saleOrderLines.get(purchaseOrderLine))
 		{
 			salesOrderLine.setLink_OrderLine(purchaseOrderLine);
 			InterfaceWrapperHelper.save(salesOrderLine);
+			salesOrdersToBeClosed.add(OrderId.ofRepoId(salesOrderLine.getC_Order_ID()));
+		}
+
+		if(PurchaseTypeEnum.MEDIATED.equals(purchaseType))
+		{
+			salesOrdersToBeClosed.forEach(orderBL::closeOrder);
```
Comment (metas-ts) [inline backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/order/createFrom/po_from_so/impl/CreatePOLineFromSOLinesAggregator.java:188]: Please do a Loggabble.withLogger when closing the order

File: backend/de.metas.business/src/main/java/de/metas/order/IOrderBL.java [hunk 2]
```diff
@@ -122,15 +120,15 @@
 	/**
 	 * Set Target Sales Document Type.
 	 * This method is also setting IsSOTrx to true.
-	 *
-	 * @param soDocSubType sales DocSubType
 	 */
-	void setDocTypeTargetId(I_C_Order order, String soDocSubType);
+	void setDocTypeTargetId(I_C_Order order);
+
+	void setPODocTypeTargetId(I_C_Order order, String soDocSubType);
```
Comment (metas-ts) [inline backend/de.metas.business/src/main/java/de/metas/order/IOrderBL.java:126]:
```suggestion
	void setPODocTypeTargetId(I_C_Order order, String poDocSubType);
```

File: backend/de.metas.business/src/main/java/de/metas/order/IOrderBL.java [hunk 3]
```diff
@@ -122,15 +120,15 @@
...
 
 	/**
 	 * Sets Target Document Type based on {@link I_C_Order#isSOTrx()} (Standard Order or PO)
 	 */
-	void setDocTypeTargetId(I_C_Order order);
+	void setSODocTypeTargetId(I_C_Order order, final String docSubType);
```
Comment (metas-ts) [inline backend/de.metas.business/src/main/java/de/metas/order/IOrderBL.java:131]:
```suggestion
	void setSODocTypeTargetId(I_C_Order order, final String soDocSubType);
```

File: backend/de.metas.business/src/main/java/de/metas/order/IOrderBL.java [hunk 4]
```diff
@@ -122,15 +120,15 @@
 	/**
 	 * Set Target Sales Document Type.
 	 * This method is also setting IsSOTrx to true.
-	 *
-	 * @param soDocSubType sales DocSubType
 	 */
-	void setDocTypeTargetId(I_C_Order order, String soDocSubType);
+	void setDocTypeTargetId(I_C_Order order);
```
Comment (metas-ts) [inline backend/de.metas.business/src/main/java/de/metas/order/IOrderBL.java:124]: please correct the javadoc.
Also, wdyt about renaming the metod to `setDefaultDocTypeTargetId()`

File: backend/de.metas.business/src/main/java/de/metas/order/impl/OrderBL.java [hunk 5]
```diff
@@ -343,8 +349,36 @@ public void setDocTypeTargetId(final I_C_Order order)
...
+			logger.error("No POO found for {}", docTypeQuery);
+		}
+		else
+		{
+			logger.debug("(PO) - {}", docTypeId);
+			setDocTypeTargetIdAndUpdateDescription(order, docTypeId);
+		}
+	}
+
+	public void setSODocTypeTargetId(final I_C_Order order, final String soDocSubType)
```
Comment (metas-ts) [inline backend/de.metas.business/src/main/java/de/metas/order/impl/OrderBL.java:381]: make setSODocTypeTargetId similar to  setPODocTypeTargetId in that it asserts that `order.isSOTrx()` is already correct.
then please set order.isSOTrx(true) before setSODocTypeTargetId is called

**Review by metas-ts — CHANGES_REQUESTED** [review]

File: backend/de.metas.cucumber/src/test/resources/de/metas/cucumber/features/salesOrder.feature [hunk 6]
```diff
@@ -30,3 +33,51 @@ Feature: sales order
...
+      | C_BPartner_ID.Identifier | C_Order_ID.Identifier | PurchaseType |
+      | vendor_1                 | o_2                   | Mediated     |
+    Then the order is created:
+      | Link_Order_ID.Identifier | IsSOTrx | DocBaseType | DocSubType |
+      | o_2                      | false   | POO         | MED        |
+    And the mediated purchase order linked to order 'o_2' has lines:
+      | QtyOrdered | LineNetAmt | M_Product_ID.Identifier |
+      | 10         | 100        | p_2                     |
+    And the sales order identified by 'o_2' is closed
+    And the shipment schedule identified by s_ol_2 is processed
```
Comment (metas-ts) [inline backend/de.metas.cucumber/src/test/resources/de/metas/cucumber/features/salesOrder.feature:83]: The shipment-schedule's IsProcessed flag is updated asynchronously via `UpdateInvalidShipmentSchedulesWorkpackageProcessor`, so i think this is way too flacky.
Please wait for some time. See `Then after not more than 10s, M_ShipmentSchedules are found:` in line 33 of this file.

**Review by metas-ts — CHANGES_REQUESTED** [review]

File: backend/de.metas.cucumber/src/test/resources/de/metas/cucumber/features/salesOrder.feature (hunk 6 above)
Comment (fp-ps) [reply backend/de.metas.cucumber/src/test/resources/de/metas/cucumber/features/salesOrder.feature:83]: right, on it

**Review by metas-ts — APPROVED** [review]

**Pull request merged**